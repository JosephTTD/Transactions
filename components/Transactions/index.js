import React from "react";
import { useState, useEffect } from "react";
import Container from "./style";
import Card from "../Card";
import Transaction from "../Transaction";

function Transactions() {
  const [data, setData] = useState(null);
  const [balance , setBalance] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  // generates a random hexcode, used for the Transaction 'crest' prop, just a stylistic choice
  function randomColor() {
    const rangeSize = 10;
    const parts = [
      Math.floor(Math.random() * 256),
      Math.floor(Math.random() * rangeSize),
      Math.floor(Math.random() * rangeSize) + 256 - rangeSize,
    ].sort((a, b) => Math.random() < 0.5);

    return parts.map((p) => ("f" + p.toString(14)).slice(-2)).join("");
  }

  useEffect(() => {
    // API request
    const getTransactions = async () => {
      const url = "http://www.mocky.io/v2/5c62e7c33000004a00019b05";
      try {
        const response = await fetch(url, {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
          },
        });
        if (!response.ok) {
          throw new Error(`Error: ${response.status}`);
        }
        let data = await response.json();
        let transactions = data.transactions;

        // sorting transactions by amount value and slicing the first 10
        const sortAmount = transactions.sort(
            (a, b) => parseFloat(a.amount.value) - parseFloat(b.amount.value)
        ).slice(0, transactions.length > 10 ? 10 : transactions.length);

        // then sorting by date / most recent first
        const sortByDate = sortAmount.sort(function (a, b) {
          return new Date(b.date) - new Date(a.date);
        });
        
        setBalance(data.balance);
        setData(sortByDate);
        setError(null);
      } catch (err) {
        setData(null);
        setBalance(null);
        setError(err.message);
      } finally {
        setLoading(false);
      }
    };
    getTransactions();
  }, []);

  // loader visual before data is set
  if (loading === true && !data?.length) {
    return (
      <Container loader>
        <img src="images/loader.svg" />
      </Container>
    );
  }

  if (error) {
    return (
      <Container loader>
        <p>{error}</p>
      </Container>
    );
  }

  return (
    <Container>
      <section>
        <Card amount={balance?.amount} name="MR J DOE" doe="09/26" />
      </section>

      <h3>Transactions</h3>

      <section>
        {data &&
          data.map((value, index) => {
            return (
              <Transaction
                key={value.id || index}
                name={value.description}
                value={value.amount.value}
                currency={value.amount.currency_iso}
                category={value.category_title}
                crest={"#" + randomColor()}
              ></Transaction>
            );
          })}
      </section>
    </Container>
  );
}

export default Transactions;
