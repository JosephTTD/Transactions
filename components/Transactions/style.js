import styled, { css } from "styled-components";

const Container = styled.div`
  background: white;
  color: black;
  border-radius: 12px;
  padding: 30px 25px;
  width: 320px;
  min-height: 75vh;
  height: auto;
  margin: 0 auto;
  position: relative;

  section:nth-child(2n + 3) {
    height: 450px;
    overflow-y: auto;
  }

  h3 {
    font-size: 15px;
  }

  ${(props) =>
    props.loader &&
    css`
      @keyframes spin {
        from {
          transform: rotate(0deg);
        }
        to {
          transform: rotate(359deg);
        }
      }

      img {
        animation: spin 1s linear infinite;
      }

      display: flex;
      align-items: center;
      justify-content: center;
    `};

  h3 {
    font-weight: 600;
  }
`;

export default Container;
