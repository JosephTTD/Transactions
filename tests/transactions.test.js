import React from "react";
import * as renderer from "react-test-renderer";
import Transactions from "../components/Transactions";

const component = renderer.create(<Transactions />);

describe("Transactions", () => {
  // snapshot tests for Transactions component

  it("should render the Transactions component correctly", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
