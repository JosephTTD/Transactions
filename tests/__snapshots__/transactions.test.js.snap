// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions should render the Transactions component correctly 1`] = `
<div
  className="sc-bczRLJ grjOHx"
>
  <img
    src="images/loader.svg"
  />
</div>
`;
